src/java/org/apache/cassandra/index/sai/disk/v1/V1OnDiskFormat.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.cassandra.index.sai.memory.RowMapping;
+import org.apache.cassandra.index.sai.utils.NamedMemoryLimiter;
+import org.apache.cassandra.index.sai.utils.SAICodecUtils;
+import org.apache.cassandra.schema.CompressionParams;
+import org.apache.lucene.store.IndexInput;
+
+import static org.apache.cassandra.index.sai.StorageAttachedIndex.SEGMENT_BUILD_MEMORY_LIMITER;
+import static org.apache.cassandra.utils.FBUtilities.prettyPrintMemory;
+
+public class V1OnDiskFormat implements OnDiskFormat
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private static final EnumSet<IndexComponent.Type> LITERAL_COMPONENT_TYPES = EnumSet.of(IndexComponent.Type.COLUMN_COMPLETION_MARKER,
+                                                                                           IndexComponent.Type.META,
+                                                                                           IndexComponent.Type.TERMS_DATA,
+                                                                                           IndexComponent.Type.POSTING_LISTS);
+    private static final EnumSet<IndexComponent.Type> NUMERIC_COMPONENT_TYPES = EnumSet.of(IndexComponent.Type.COLUMN_COMPLETION_MARKER,
+                                                                                           IndexComponent.Type.META,
+                                                                                           IndexComponent.Type.KD_TREE,
+                                                                                           IndexComponent.Type.KD_TREE_POSTING_LISTS);
+
+    public static final V1OnDiskFormat instance = new V1OnDiskFormat();
+
+    private V1OnDiskFormat()
+    {}
+
+    @Override
+    public boolean isGroupIndexComplete(IndexDescriptor indexDescriptor)
+    {
+        return indexDescriptor.registerSSTable().hasComponent(IndexComponent.GROUP_COMPLETION_MARKER);
+    }
+
+    @Override
+    public boolean isColumnIndexComplete(IndexDescriptor indexDescriptor, IndexContext indexContext)
+    {
+        indexDescriptor.registerSSTable().registerIndex(indexContext);
+        return indexDescriptor.hasComponent(IndexComponent.GROUP_COMPLETION_MARKER) &&
+               indexDescriptor.hasComponent(IndexComponent.create(IndexComponent.Type.COLUMN_COMPLETION_MARKER, indexContext.getIndexName()));
+    }
+
+    @Override
+    public boolean isCompletionMarker(IndexComponent indexComponent)
+    {
+        return indexComponent.type == IndexComponent.Type.GROUP_COMPLETION_MARKER ||
+               indexComponent.type == IndexComponent.Type.COLUMN_COMPLETION_MARKER;
+    }
+
+    @Override
+    public boolean isEncryptable(IndexComponent indexComponent)
+    {
+        return indexComponent.type == IndexComponent.Type.TERMS_DATA;
+    }
+
+    @Override
+    public PerSSTableComponentsWriter createPerSSTableComponentsWriter(boolean perColumnOnly,
```

Comment:
What is the meaning of `perColumnOnly`? Why if `perColumnOnly == true` it returns a noop writer?

Reply:
Yes, another badly named field. The point here is that when we build the indexes for a SSTable we only need to build the per-SSTable files once. The first index builder kicked off does this and the others use the noop builder instead. I'll rename this to something more understandable.